src/simulator/halo_simulator.py

```diff
@@ -307,6 +311,37 @@ def _form_venn_diagram_regions(
 
         return regions
 
+    def _add_dp_noise_to_primitive_regions(
+        self,
+        occupied_primitive_regions: Dict[int, List],
+        num_all_primitive_regions: int,
+        noiser: EstimateNoiserBase,
+        noise_event: NoisingEvent,
```

Comment:
I made the noiser and the noise event as part of the input because:
1. more general and decoupled
2. easy to test the function

but I am open to suggestions.

Reply:
Offline discussion with @matthewclegg:
In order to create a noise event right beside code of adding noise, we decide to move the creation of the noise event inside the function. In addition, since the noiser is highly related to the noise event, we also move the nosier inside the function.

Reply:
Done.